File: app/services/rbac/gated_types.py
"""
Gated media types: a type every entry of which carries one content label.

A media type named here is a GATED TYPE. Every entry of it carries its
required label, so a session whose mode lacks that label sees none of the
type's entries - and, through the shared-record rule in
`shared_visibility.py`, none of the people, vocabulary values or other
records that exist only for it.

The map is keyed on the hyphenated media-type key (`spec.owner_type`, the
form every `media_type` column and permission uses) and names a
`content_label.key`, not an id: ids are minted per database, and the key is
what a seed, a migration and a test can all spell.

A viewer CAN SEE a gated type when the type's required label is not in the
viewer's hidden label set (`enforcement.hidden_label_ids`). An ungated type -
every type not named here - is always seeable on this axis; whether the
viewer holds `media_type.<key>` is the role axis, asked separately by
`Viewer.has()`.
"""

from typing import Iterable, Optional
from uuid import UUID

from sqlalchemy.orm import Session

from app import models
from app.services.rbac.enforcement import hidden_label_ids

# media-type key -> content-label key. The label is a system label: created
# by its migration and by the lifespan seed
# (gated_labels.ensure_system_labels), attached to every entry of the type on
# every write path (app/services/domain/gated_labels.py), and granted to no
# mode but `unrestricted` - see ensure_access_mode_seed, which keeps it off
# the other all-labels mode. A new gated type adds its entry here and its row
# to gated_labels.SYSTEM_LABELS.
REQUIRED_LABEL_FOR_TYPE: dict[str, str] = {
    "h-comic": "h-comic",
    "h-game": "h-game",
}


def required_label_keys() -> frozenset[str]:
    """Every content-label key some gated type requires."""
    return frozenset(REQUIRED_LABEL_FOR_TYPE.values())


def visible_gated_types(db: Session, viewer: Optional[object]) -> list[str]:
    """
    The gated types this viewer may see, sorted.

    What /api/auth/me publishes so the SPA can decide whether to offer a gated
    type's navigation at all. Only the SEEABLE ones are named: a session that
    cannot see a type is not told that it exists.
    """
    if not REQUIRED_LABEL_FOR_TYPE:
        return []
    if viewer is None:
        return sorted(REQUIRED_LABEL_FOR_TYPE)
    hidden = hidden_gated_types(db, hidden_label_ids(db, viewer))
    return sorted(t for t in REQUIRED_LABEL_FOR_TYPE if t not in hidden)


def gated_types() -> frozenset[str]:
    """Every media type that names a required label."""
    return frozenset(REQUIRED_LABEL_FOR_TYPE)


def hidden_gated_types(db: Session, hidden: Iterable[UUID]) -> frozenset[str]:
    """
    The gated types whose required label is in `hidden`.

    `hidden` is `enforcement.hidden_label_ids(db, viewer)`, passed in rather
    than recomputed so a caller asking both questions pays for the label
    query once. A required label that does not exist as a row is not hidden:
    no mode can lack a label nobody has created, and the type is then gated
    on nothing.
    """
    hidden = set(hidden)
    if not REQUIRED_LABEL_FOR_TYPE or not hidden:
        return frozenset()
    ids_by_key = dict(
        db.query(models.ContentLabel.key, models.ContentLabel.system_id)
        .filter(models.ContentLabel.key.in_(set(REQUIRED_LABEL_FOR_TYPE.values())))
        .all()
    )
    return frozenset(
        media_type
        for media_type, label_key in REQUIRED_LABEL_FOR_TYPE.items()
        if ids_by_key.get(label_key) in hidden
    )


def can_see_gated_type(db: Session, viewer: Optional[object], media_type: str) -> bool:
    """
    Whether `viewer` may see gated type `media_type`.

    True for an ungated type, and for a `None` viewer - "not a request", as
    everywhere in `enforcement.py`.
    """
    if viewer is None or media_type not in REQUIRED_LABEL_FOR_TYPE:
        return True
    return media_type not in hidden_gated_types(db, hidden_label_ids(db, viewer))


# ---------------------------------------------------------------------------
# What a session that cannot see a gated type is not told
# ---------------------------------------------------------------------------
# Hiding a gated type's entries is not enough: a vocabulary, a role or a
# dropdown that exists ONLY for the type names it just as surely. Each helper
# below answers one "does this exist only for types the viewer cannot see?"
# question from a declaration the code already makes - a role's media types,
# a tag field's media types, a franchise type's media types - so a second
# gated type needs no edit here.


def unseeable_gated_types(db: Session, viewer: Optional[object]) -> frozenset[str]:
    """The gated types `viewer` may not see. Empty for a None viewer."""
    if viewer is None or not REQUIRED_LABEL_FOR_TYPE:
        return frozenset()
    return hidden_gated_types(db, hidden_label_ids(db, viewer))


def only_for(media_types, hidden: frozenset[str]) -> bool:
    """True when every one of `media_types` is hidden - a non-empty set only."""
    media_types = set(media_types)
    return bool(media_types) and media_types <= set(hidden)


def gated_tag_categories() -> dict[str, frozenset[str]]:
    """
    {option category: the media types it serves}, for every category whose
    tag fields serve GATED types only - the h-comic genres, not the shared
    Official Source vocabulary. Such a category is itself a connection of
    every value in it (shared_visibility.py), so a value an admin creates with
    no scope row is still hidden with its type.
    """
    from app.utils.credit_roles import TAG_FIELDS

    served: dict[str, set[str]] = {}
    for field in TAG_FIELDS.values():
        served.setdefault(field.category, set()).update(field.media_types)
    gated = gated_types()
    return {
        category: frozenset(types)
        for category, types in served.items()
        if types and types <= gated
    }


def hidden_person_roles(hidden: frozenset[str]) -> frozenset[str]:
    """Person roles every legal scope of which is a hidden gated type."""
    from app.utils.credit_roles import CREDIT_ROLES

    return frozenset(
        key
        for key, role in CREDIT_ROLES.items()
        if role.target == "person" and only_for(role.media_types, hidden)
    )


def hidden_option_categories(hidden: frozenset[str]) -> frozenset[str]:
    """Option categories that serve hidden gated types only."""
    return frozenset(
        category
        for category, types in gated_tag_categories().items()
        if types <= hidden
    )


def hidden_franchise_types(hidden: frozenset[str]) -> frozenset[str]:
    """Franchise types stamped only for hidden gated types (H-Comic)."""
    from app.services.domain.hierarchy import FRANCHISE_TYPE_FOR

    served: dict[str, set[str]] = {}
    for media_type, franchise_type in FRANCHISE_TYPE_FOR.items():
        served.setdefault(str(getattr(franchise_type, "value", franchise_type)), set()).add(
            media_type
        )
    return frozenset(
        franchise_type
        for franchise_type, types in served.items()
        if only_for(types, hidden)
    )

File: frontend/src/lib/gatedTypes.js
// Frontend: the gated media types, and the one question every surface asks
// about them.
//
// A gated type is a media type whose every entry carries a required content
// label (REQUIRED_LABEL_FOR_TYPE in app/services/rbac/gated_types.py) - today
// `h-comic` and `hentai`. A session whose access mode lacks that label is not told
// the type exists: the server withholds its entries, its vocabularies and its
// constants, and `/api/auth/me` names in `visible_gated_types` only the gated
// types the session MAY see.
//
// So the SPA never decides visibility itself. It asks `canSeeGatedType`, which
// reads that list, and every surface that could name a gated type - the nav,
// both routes, the type pickers, the search tabs, the statistics chips - asks
// through here rather than testing `"h-comic"` inline. Hiding in the UI is
// cosmetic: the server has already refused what the viewer may not read.

// Every gated media type the frontend knows how to render. A type that is not
// listed here is ungated and always visible.
export const GATED_TYPES = Object.freeze(["h-comic", "hentai"]);

// Franchise types stamped only for a gated media type (FRANCHISE_TYPE_FOR on
// the backend). A franchise-type picker offers one only when its media type
// is visible.
export const GATED_FRANCHISE_TYPES = Object.freeze({
  "H-Comic": "h-comic",
  Hentai: "hentai",
});

const GATED = new Set(GATED_TYPES);

// Franchise FAMILIES (FRANCHISE_FAMILY_FOR_TYPE in app/utils/constants.py):
// which franchise types may share one franchise. An h-comic and its hentai
// adaptation share one the way a manga and its anime do, so H-Comic and
// Hentai are one family; a type not listed is mainstream. The server refuses
// an entry written into a franchise of another family, so a gated type's
// franchise picker offers only its own family's franchises.
export const FRANCHISE_FAMILY_FOR_TYPE = Object.freeze({
  "H-Comic": "h-comic",
  Hentai: "h-comic",
});

/**
 * True when a franchise whose `franchise_type` is this comma-joined list
 * ("H-Comic, Hentai") belongs to `family`.
 */
export function inFranchiseFamily(franchiseType, family) {
  return String(franchiseType || "")
    .split(",")
    .map((t) => t.trim())
    .some((t) => FRANCHISE_FAMILY_FOR_TYPE[t] === family);
}

/** True when `type` is one of the gated media types. */
export function isGatedType(type) {
  return GATED.has(type);
}

/**
 * May this session see media type `type`?
 *
 * `auth` is the AuthContext value (or anything carrying
 * `visibleGatedTypes`). An ungated type is always visible; a gated one only
 * when the server named it. A missing list means none - a session whose
 * `/api/auth/me` has not answered yet is shown nothing gated rather than
 * everything.
 */
export function canSeeGatedType(auth, type) {
  if (!GATED.has(type)) return true;
  const visible = auth?.visibleGatedTypes;
  return Array.isArray(visible) && visible.includes(type);
}

/** `types` without the gated ones this session may not see. */
export function visibleMediaTypes(auth, types) {
  return types.filter((type) => canSeeGatedType(auth, type));
}

/**
 * The same filter over any list of objects naming a media type, e.g. the
 * `{ key, label }` tabs of the search page or `{ value, label }` pickers.
 * `keyOf` picks the type out of one item.
 */
export function visibleByType(auth, items, keyOf = (item) => item.key) {
  return items.filter((item) => canSeeGatedType(auth, keyOf(item)));
}

/** Franchise types, without those stamped only for a hidden gated type. */
export function visibleFranchiseTypes(auth, franchiseTypes) {
  return franchiseTypes.filter((ft) => {
    const gated = GATED_FRANCHISE_TYPES[ft];
    return !gated || canSeeGatedType(auth, gated);
  });
}

// The content label each gated type requires (REQUIRED_LABEL_FOR_TYPE on the
// backend). Every entry of the type carries it, and so does every franchise
// of its franchise type; the content-label endpoints refuse (422) a new set
// that drops it. So the label pickers lock it on, and the savers add it back
// rather than let a save the admin never touched fail.
export const REQUIRED_LABEL_FOR_TYPE = Object.freeze({
  "h-comic": "h-comic",
  hentai: "hentai",
});

/** The labels an entry of `mediaType` must carry: [] for an ungated type. */
export function requiredLabelsForType(mediaType) {
  const label = REQUIRED_LABEL_FOR_TYPE[mediaType];
  return label ? [label] : [];
}

/**
 * The labels a franchise of this `franchise_type` must carry. The column is a
 * comma-joined list ("ACG, Game"), and a franchise that names a gated
 * franchise type anywhere in it carries that type's label.
 */
export function requiredLabelsForFranchiseType(franchiseType) {
  const named = String(franchiseType || "")
    .split(",")
    .map((t) => t.trim())
    .filter(Boolean);
  return [
    ...new Set(
      named.flatMap((ft) =>
        GATED_FRANCHISE_TYPES[ft] ? requiredLabelsForType(GATED_FRANCHISE_TYPES[ft]) : []
      )
    ),
  ];
}
